js/src/common/components/Modal.tsx

```diff
@@ -1,33 +1,42 @@
 import Component from '../Component';
-import Alert from './Alert';
+import Alert, { AlertAttrs } from './Alert';
 import Button from './Button';
 
+import type Mithril from 'mithril';
+import type ModalManagerState from '../states/ModalManagerState';
+import type RequestError from '../utils/RequestError';
+import type ModalManager from './ModalManager';
+
+interface IInternalModalAttrs {
+  state: ModalManagerState;
+  animateShow: ModalManager['animateShow'];
+  animateHide: ModalManager['animateHide'];
+}
+
 /**
  * The `Modal` component displays a modal dialog, wrapped in a form. Subclasses
  * should implement the `className`, `title`, and `content` methods.
- *
- * @abstract
  */
-export default class Modal extends Component {
+export default abstract class Modal<ModalAttrs extends Record<string, unknown> = {}> extends Component<ModalAttrs & IInternalModalAttrs> {
```

Comment:
What's the reasoning behind splitting the internal & other attrs? Don't all methods use `ModalAttrs & IInternalModalAttrs` anyways (both in this class & parent Component)?

This just seems like it overcomplicates the code for no reason, I'd rather just have a `ModalAttrs` interface that extends `ComponentAttrs`

Reply:
> I'd rather just have a `ModalAttrs` interface that extends `ComponentAttrs`

`ComponentAttrs` includes native HTML attributes, which don't apply in this situation. Extending with it would add typing for `className`, `id`, etc, which aren't inherently valid Modal attributes.

Reply:
Ah, I see. Can't we just have a ModalAttrs interface that contains teh `IInternalModalAttrs` attrs and extends `Record<string, unknown>` so we don't need to see `ModalAttrs & IInternalModalAttrs` everywhere? Or nah?

Reply:
I don't like the idea of extending with `Record<string, unknown>` as that allows props that haven't actually been typed on the modal.

https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgJICUD2mzIgD0hABMBnZdCBTKYgHlLClAHMAaZAVxAGsRMA7iAB8yAN4BYAFDJZyALYBPAApRMABwBcyRsxAtpAX2nTQkWIhSpVG8gSJk0WHOOlyFigPJgAFtBtayCCc8gBG0EYmUtQgjMj42tZq6uQAvK4yckoB2gBEPsAAhLlsbllevv7J2gCMpZmyvsCkqKQAcjgAKorqEMR5oZy4zTpgwAA248iISOqQxMihisidAMolkVJAA

Reply:
Doesn't doing `ModalAttrs extends Record<string, unknown> = {}` achieve the same thing? 🤔

Reply:
Oooooooooops, yes it does. I meant to remove that.